### PR TITLE
Show days in ETA if needed

### PR DESCRIPTION
**Pull request by mvglasow**
As discussed in #1038. This PR adds the following functionality:

If ETA >= 86220 s (which rounds up to 86400 s, or 24 h, for display purposes), it is displayed as:

    0%  0/1  (40.3 MiB of 263.59 GiB)   357 KiB/s    2d 20:25:00 ETA

ETAs below that threshold will get displayed as before.

Since this is my first time messing with any OCaml code, you might want to look at the syntax. Lines 496–501 are truly new, the others just had indentation added. In particular, I am not sure if the parentheses around the tuple members in 497/498 are correct, or if they’re needed at all.

I have run a brief test with ETAs both below and above the threshold, and it builds and works as intended.

### CODE REVIEW

File: src/uicommon.ml
Comment (tleedjarv) [inline src/uicommon.ml:500]: Overall looks correct but I propose to simplify a little since now you don't need the tuple at all. At the same time, I'd consider reducing the precision even more.
Something like this (not tested):
```suggestion
    let u = t + 1799 in
    Printf.sprintf "%dd %02d:%02d:00" (u / 86400) ((u mod 86400) / 3600)
      (((u mod 3600) / 1800) * 30)
```
This makes the precision 30 min (currently rounded fully up, but can be adjusted by changing the constant `1799`). This could be anything, really.

Alternatively, you can still express all variables separately, but still don't need the tuple:
```ocaml
    let u = t + 1799 in
    let d = u / 86400 in
    let h = (u mod 86400) / 3600 in
    let m = ((u mod 3600) / 1800) * 30 in
    Printf.sprintf "%dd %02d:%02d:00" d h m
```
but I'm not sure it adds anything.

Comment (mvglasow) [reply src/uicommon.ml:500]: Fine with me, but then I would `let u = t + 1800` and make the threshold 84800 (86400 - 1800), to be consistent with the rest of the code.

Comment (tleedjarv) [reply src/uicommon.ml]: You could, but is it the best idea? Those thresholds are not directly related. You can still keep it at 86220 because that goes together with the < 24h threshold. If you make the threshold (86400 - 1800) then you cut the 24h range a bit short (meaning you'll jump from 23:30:00 to 1d 00:00:00). But that's ok too, all of this is completely arbitrary; there is nothing specifying that it should be 5 min precision up to 24h.

Comment (tleedjarv) [reply src/uicommon.ml]: > to be consistent with the rest of the code.

That's not making it consistent. Those thresholds are not related and you can see these differ in the code below. They are arbitrary, true, but they are not the same.

Comment (mvglasow) [reply src/uicommon.ml]: Indeed, I just went by the next lower threshold: for ETAs above roughly an hour, we add 180 s and make the threshold 3420 s.

Looking at the code again, however, I realize that for any `3420 > t >= 2640`, we effectively do the same as for `86400-ish > t >= 3420`, so we could simplify that part as well.

And below that, the thresholds are “true” rather than rounded ETA.

Reconsidering it all, I’d lean towards sticking with 5-minute precision up to `1d 00:00:00` or `24:00:00` (I guess the former format would be more consistent). So that would put us back to 86220...

Comment (mvglasow) [reply src/uicommon.ml]: In short, no further commits on top of your suggested improvements. The rest of the commit history in this PR cancels itself out.

File: src/uicommon.ml
Comment (gdt) [inline src/uicommon.ml:498]: I am really not a fan of displaying a :00 for minutes, if I follow that. 

I am ok with us deciding that these strings are only for humans.  (If they are for computers, they should just be in seconds, the SI base unit :-)    If so, we could adopt a policy that if there are days, we should only days and hours, and if hours, only hours and minutes, and so on.   But I think showing a 0 in a field when it is not, is not ok.

Overall I lean to simpler, in code but especially in presentation to humans.

**Comment by tleedjarv** [comment]
@mvglasow maybe that's just how GitHub's suggestions work but you don't have to assign any co-authorship to me and can simply squash the commits.

**Review by tleedjarv — APPROVED** [review]
I think it looks ok.
Some primitive testing shows:
85919 is displayed as 23:50:00
86040 is displayed as 23:55:00
86218 is displayed as 23:55:00
86219 is displayed as 23:55:00
86220 is displayed as 1d 00:00:00
86221 is displayed as 1d 00:00:00
86399 is displayed as 1d 00:00:00
86400 is displayed as 1d 00:00:00
86401 is displayed as 1d 00:30:00
88199 is displayed as 1d 00:30:00
88200 is displayed as 1d 00:30:00
88201 is displayed as 1d 01:00:00
187800 is displayed as 2d 04:30:00

File: src/uicommon.ml
Comment (tleedjarv) [reply src/uicommon.ml:498]: (It's for the seconds in this case.)

I'm not sure I agree with you on this one. It sounds like form over function. How do you avoid ambiguity? You'd have to switch away from positional display then, right?

But this discussion should not extend the scope of this PR. I think this discussion instead goes well together with a discussion of the precision of time display (also separate from this PR).

**Comment by gdt** [comment]
This PR is about changing large numbers of hours into days and hours.  I am fine with that.    I do not want to have a change to display rounded times in the same PR; that's a different discussion that we don't even have an issue about :-)

File: src/uicommon.ml
Comment (mvglasow) [reply src/uicommon.ml:498]: > showing a 0 in a field when it is not, is not ok.

For seconds this is what is already happening for any ETA above 2 minutes. Between 1–2 minutes ETA is rounded to the next 30 seconds, and below a minute the exact number of seconds is shown. The only change regarding rounding is that above 24 hours we further reduce precision from 5 to 30 minutes.

I don’t have a clear position on whether precision for ETAs above 24 h should be 5 m or 30 m. I’d suggest merging this PR for now – I’ll leave it up to you guys whether or not to make any changes to precision for ETA > 24 h before merging.

If there is a need to discuss what rounding we should apply when, or whether to display seconds when they are always zero, I would suggest doing so in a separate ticket. For seconds my only requirement is that I don’t want any ambiguity between `hh:mm` and `mm:ss`, even when I just have a single, static value such as a single log line.

**Comment by gdt** [comment]
Agreed that varying between hh:mm and mm:ss is not ok.

I feel that the rounding we already have is overly aggressive.   Agreed that fixing that is separate.   But I'd like to not pile on and make it worse.   If we already have "round to 5 minutes" for times that are most of a day, flipping to 30 minutes at a day seems too much (4%).  (Yes, it might be that the existing breakpoint for 5m is also too much, but we're not changing that.)   If this just doesn't change rounding from how it is (leaving rounding at 5 for multiday), I'm fine with it.

**Comment by mvglasow** [comment]
I’m fine with keeping precision at 5 minutes for ETAs beyond 24 h for now.

**Comment by gdt** [comment]
Great then please adjust, rebase, and make the commit message give rationale for the change.  I don't mean anything super extensive, but it's a wee bit too brief as is

**Comment by mvglasow** [comment]
Just committed. Github should allow a squash merge, which will squash all the commits from my branch into a single one – should be the easiest way. If not, give me a heads-up and I’ll see if I can rewrite my branch history.

**Comment by gdt** [comment]
I do not expect github squash to do a good job with a merged commit message turning into the right message for posterity, and we do tend to have merge commits, so I think you can just do a rebase -i, p/s/s and force-push.

**Comment by mvglasow** [comment]
OK. `git rebase` didn’t do anything as master has not changed, so I did `git rest master`, then re-committed the differences and force-pushed them to my branch. Commit message is the same as the subject of this PR, which should accurately describe what we’re doing (and, afaik, is also what Github uses as the commit message for a squash commit).

**Comment by gdt** [comment]
The commit message is only really understandable once you have the context.  Let me think about how to say enough to orient people who have no idea.    It's allowed to have more lines after the 54-byte summary :-)   I'll wait a day or so for anybody else to comment.

(I meant `git rebase -i origin/master` which will give you a chance to specify how you want things in terms of squashing, fixups and reordering, vs rebase which will not.)   Thanks for pointing out about github and squash, but I basically think it's a bug we are github and don't want to rely on such behavior anyway, trying to be closer to pure git when reasonable.

**Comment by gdt** [comment]
I'm not really 100% happy with the entire time display process, but on re-reading am ok with this being an improvement.   Sorry for it taking so long; I often have unison paged out pretty much entirely.

**Pull request merged**
